### DISCUSSION TITLE
Linux Key Handler

### DISCUSSION THREAD

**Discussion by kanjoe24** [rdkcentral/ut-core#148] [Ideas]
### Overview

This application enables generating and emitting virtual input events to the Linux input subsystem via the uinput interface. It uses a user-specified profile YAML file to configure key-event mappings of event type, key code, and value via command-line arguments.

### Features:

- Emit Input Events: Generate and send input events programmatically.
- Profile Support: Leverage a YAML file for defining key mappings and event types.
- Manual Input: Specify event type, code, and value via command-line arguments.
- UInput Integration: Use Linux’s /dev/uinput to create and manage virtual devices.

### Prerequisites

- Linux system with uinput module enabled.
- Necessary permissions to write to /dev/uinput.
- A valid profile YAML file.

### Command-Line Usage
**Basic Syntax**
```
./ut_core_key_input -p input-event-codes-5.15.yaml -c KEY_SPACE -t EV_KEY -v 1

```

**Options:**

- -p <profile_file> : Path to the profile YAML file defining event types and key mappings. Required.
- -t <event_type> : Type of the event (e.g., EV_KEY). Required.
- -c <event_code> : Event code to emit (e.g., KEY_A). Required.
- -v <value> : Event value to emit: 1,0. Required.

### How It Works

1. 	Profile Loading:
        Reads the profile YAML file using the ut_kvp library.
        Extracts event type and key mappings from yaml file using KVP.
2. 	UInput Setup:
        Opens /dev/uinput with write permissions.
        Configures the virtual device with UI_DEV_SETUP and UI_DEV_CREATE.
3. 	Event Emission:
        Uses the emit function to send key events to the virtual device.
        Automatically includes SYN_REPORT to signal event completion.
4. Device Teardown:
    Cleans up kvp resources and destroys the virtual input device with UI_DEV_DESTROY.

**Comment by Ulrond** [comment]
you should remove -> `./<application_name> and make it ut_core_key_input

it's not going to - validate the file
-> Configures the virtual device with UI_DEV_SETUP and UI_DEV_CREATE. ?? not seen that in teh ode
-> whys it talking about virtual device? Uses the emit function to send key events to the virtual device.

-> does it ? -> Automatically includes SYN_REPORT to signal event completion.
-> Cleans up kvp resources and destroys the virtual input device with UI_DEV_DESTROY.??

remove 

Limitations
Requires the uinput kernel module to be loaded.
Assumes the YAML profile file is correctly formatted and contains valid key mappings.

**Reply by kanjoe24** [reply]
We do use UI_DEV_SETUP, UI_DEV_CREATE and UI_DEV_DESTROY . https://github.com/rdkcentral/ut-core/blob/a55f2c0994575c260e01cb3d6d301bd08020c7f4/keyhandler/keyhandler.c#L184

yes, we are sending th SYN_REPORT after every emit : https://github.com/rdkcentral/ut-core/blob/a55f2c0994575c260e01cb3d6d301bd08020c7f4/keyhandler/keyhandler.c#L53
